qiskit/compiler/transpiler.py

```diff
@@ -5,46 +5,322 @@
 # This source code is licensed under the Apache License, Version 2.0 found in
 # the LICENSE.txt file in the root directory of this source tree.
 
-"""Circuit transpile function """
-import logging
+"""Circuit transpile function"""
+import warnings
 
-from qiskit import transpiler
-from qiskit.mapper import Layout
+from qiskit.mapper import Layout, CouplingMap
+from qiskit.tools.parallel import parallel_map
+from qiskit.transpiler.preset_passmanagers import (default_pass_manager_simulator,
+                                                   default_pass_manager)
+from qiskit.compiler.transpile_config import TranspileConfig
 
 
-logger = logging.getLogger(__name__)
+def transpile(circuits,
+              basis_gates=None, coupling_map=None, backend_properties=None,
+              initial_layout=None, seed_transpiler=None, optimization_level=None,
+              transpile_config=None, backend=None,
+              seed_mapper=None, pass_manager=None):
+    """transpile one or more circuits, according to some desired
+    transpilation targets.
 
+    All arguments may be given as either singleton or list. In case of list,
+    the length must be equal to the number of circuits being transpiled.
 
-def transpile(circuits, transpile_config=None):
```

Comment:
With this PR, `TranspileConfig` is used only internally which undermines the point of `TranspilerConfig`.

Reply:
This was decided some time ago by @ajavadia, @jaygambetta, and myself.  If you use a config object in the transpiler, then the user must learn yet another object.  In addition, the signature of the `transpile` function would not elucidate the arguments that are needed in the transpilation process.  It was decided (mostly by my strong arguing) that we should not use config objects in user facing functions.  We are of course free to use them internally.

Reply:
I updated the description to reflect this. But yeah basically the config objects give some structure to terra's internal functions, but we allow the user to pass kwargs for convenience.